### ISSUE TITLE
RFE: Support for AptX LL (Low-Latency)

### ISSUE THREAD

**Issue by pasikarkkainen** [pali/libopenaptx#1]
AptX-LL (Low Latency) is meant for "Vital synchronised audio and video for gaming and multimedia applications". AptX LL has audio latency of around 32-40ms, which is much less than the other codec options currently available for Bluetooth audio. Currently supported AptX and AptX-HD codecs in libopenaptx have higher latency compared to AptX-LL.

**Comment by pali** [comment]
I do not have codec details for aptX Low Latency yet, nor any reference implementation for testing.

**Comment by pasikarkkainen** [comment]
Yep. For testing there are multiple BT receiver/transmitter dongles (with 3,5 mm analog audio connectors) out there, based on the Qualcomm CSR8670 chip, which support AptX LL. There's also the Qualcomm CSR8675 / BC8675, which also supports AptX HD in addition to Low Latency. Many of these devices can indicate the codec used with LEDs, and are OK for testing this stuff. 

I personally have one of these devices, and can help with testing, if information about LL becomes available one day and can be implemented in libopenaptx.

Examples of such external devices: Inateck BR1006, Sonru Bluetooth 5.0 Transmitter Receiver, or Avantree Audikast. There are many other devices out there aswell.

**Comment by pali** [comment]
External HW player is not enough. For testing it is needed something which generates (or receive) aptX LL and decodes it to PCM. Some equivalent of decoder.

**Comment by pali** [comment]
Also aptX LL encoded samples are needed for development.

**Comment by ValdikSS** [comment]
Can confirm that current version of libopenaptx successfully decodes aptX Low Latency-encoded audio.

**Comment by pali** [comment]
I have already experimental aptX-LL patches for pulseaudio and I will send them to mailing list after we fix issues around API and other stuff which is now more important.

**Comment by ashleysommer** [comment]
@pali any chance you can look at whether you can push your aptX-LL functionality into Pipewire a2dp?

**Comment by pali** [comment]
> @pali any chance you can look at

yes, I have looked.

> you can push your aptX-LL functionality into Pipewire a2dp?

No, I cannot. Devs&admins disallowed me to do it.

**Comment by ashleysommer** [comment]
I'm talking about Pipewire, not Pulseaudio. I thought the pipewire devs were much more receptive?

**Comment by pali** [comment]
They too, they have banned me from that project too.

**Comment by ashleysommer** [comment]
😞

**Comment by cleveHEX** [comment]
> They too, they have banned me from that project too.

May I ask you why? Your contributions are very important for me, any chance it will be working in future?

**Comment by pali** [comment]
You are asking on wrong place. And there is no chance for future.
